higher/index.ts

```diff
@@ -0,0 +1,2314 @@
+  }
+  
+  #printBadContactError() {
+    console.log("----------ERROR cannot add self as contact");
+  }
+  
+  #printBadDataPermissionsError() {
+    console.log("----------ERROR insufficient permissions for modifying data");
+  }
+  
+  #printBadGroupPermissionsError() {
+    console.log("----------ERROR insufficient permissions for modifying group");
+  }
+
+  /*
+   ********************
+   * Core interaction *
+   ********************
+   */
+
+  /**
+   * Decrypts, validates, and demultiplexes the received message to
+   * appropriate handler.
+   *
+   * TODO consider: expose ordered or unordered message reception?
+   * E.g. ordered may be simpler to program against but maybe 
+   * real-time apps would benefit from an unordered API (perhaps
+   * make this composable?).
+   *
+   * TODO make private to module instead of exporting fully.
+   *
+   * @param {{ seqID: number,
+   *           encPayload: string,
+   *           srcIdkey: string }} msg message with encrypted contents
+   */
+  async #onMessage(
+      payload: payloadType,
+      sender: string
+  ) {
+    let permissionsOK = this.#checkPermissions(payload, sender);
+    //console.log(this.demuxFunc);
+    if (this.demuxFunc === undefined) {
+      this.#printBadMessageError(payload.msgType);
+      return;
+    }
+    if (!permissionsOK) {
+      this.#printBadPermissionsError();
+      return;
+    }
+    if (!this.#validate(payload)) {
+      this.#printBadDataError();
+      return;
+    }
+    console.log("SUCCESS");
+    await this.demuxFunc(payload);
+  }
+
+  /**
+   * Resolves a list of one or more group IDs to a list of public keys.
+   *
+   * @param {string[]} ids group IDs to resolve
+   * @return {string[]}
+   *
+   * @private
+   */
+  #resolveIDs(ids: string[]): string[] {
+    let idkeys = [];
+    ids.forEach((id) => {
+      let group = this.#getGroup(id);
+      if (group !== null) {
+        if (this.#isKey(group)) {
+          idkeys.push(id);
+        } else {
+          idkeys = idkeys.concat(this.#resolveIDs(group.children));
+        }
+      }
+    });
+    return idkeys;
+  }
+  
+  /**
+   * Helper function for determining if resolveIDs has hit it's base case or not.
+   *
+   * @param {Object} group a group
+   * @returns {boolean}
+   *
+   * @private
+   */
+  #isKey(group: groupValType): boolean {
+    if (group.children) {
+      return false;
+    }
+    return true;
+  }
+
+  /*
+   ***********
+   * Devices *
+   ***********
+   */
+
+  /**
+   * Initializes a new device with a keypair and adds the public key to the
+   * server.
+   *
+   * @param {?string} linkedName human-readable name (for contacts)
+   * @param {?string} deviceName human-readable name (for self)
+   * @returns {string}
+   *
+   * @private
+   */
+  async #initDevice(
+      linkedName?: string,
+      deviceName?: string
+  ): Promise<{ idkey: string, linkedName: string }> {
+    let idkey: string = await this.core.olmWrapper.generateInitialKeys();
```

Comment:
could this be more generic? something like 
`await this.core.newId()`
so interactions with the encryption lib still go through the core lib and the layers are clearer?